internal/agent/device/spec/spec_test.go

```diff
@@ -551,13 +553,525 @@ func TestCheckOsReconciliation(t *testing.T) {
+		marshaledDesiredSpec, err := json.Marshal(renderedDesiredSpec)
+		require.NoError(err)
+		mockReadWriter.EXPECT().ReadFile(desiredPath).Return(marshaledDesiredSpec, nil)
+
+		rollbackSpec, err := createEmptyTestSpec()
+		require.NoError(err)
+		mockReadWriter.EXPECT().ReadFile(rollbackPath).Return(rollbackSpec, nil)
+
+		mockClient.EXPECT().GetRenderedDeviceSpec(ctx, gomock.Any(), gomock.Any()).Return(nil, http.StatusServiceUnavailable, specErr)
+
+		_, err = s.GetDesired(ctx, renderedVersion)
+		require.ErrorIs(err, ErrGettingDeviceSpec)
+	})
+
+	t.Run("desired spec is returned when managemnt api returns no content", func(t *testing.T) {
+		renderedVersion := "1"
+
+		renderedDesiredSpec := createRenderedTestSpec(image)
+		marshaledDesiredSpec, err := json.Marshal(renderedDesiredSpec)
+		require.NoError(err)
+		mockReadWriter.EXPECT().ReadFile(desiredPath).Return(marshaledDesiredSpec, nil)
+
+		rollbackSpec, err := createEmptyTestSpec()
+		require.NoError(err)
+		mockReadWriter.EXPECT().ReadFile(rollbackPath).Return(rollbackSpec, nil)
+
+		mockClient.EXPECT().GetRenderedDeviceSpec(ctx, gomock.Any(), gomock.Any()).Return(nil, http.StatusNoContent, nil)
+
+		specResult, err := s.GetDesired(ctx, renderedVersion)
+		require.NoError(err)
+		require.Equal(renderedDesiredSpec, specResult)
+	})
+
+	t.Run("spec from the api response has the same RenderedVersion as desired", func(t *testing.T) {
+		renderedVersion := "1"
+
+		renderedDesiredSpec := createRenderedTestSpec(image)
+		marshaledDesiredSpec, err := json.Marshal(renderedDesiredSpec)
+		require.NoError(err)
+		mockReadWriter.EXPECT().ReadFile(desiredPath).Return(marshaledDesiredSpec, nil)
+
+		rollbackSpec, err := createEmptyTestSpec()
+		require.NoError(err)
+		mockReadWriter.EXPECT().ReadFile(rollbackPath).Return(rollbackSpec, nil)
+
+		apiResponse := &v1alpha1.RenderedDeviceSpec{RenderedVersion: renderedVersion}
+		mockClient.EXPECT().GetRenderedDeviceSpec(ctx, gomock.Any(), gomock.Any()).Return(apiResponse, 200, nil)
+
+		specResult, err := s.GetDesired(ctx, renderedVersion)
+		require.NoError(err)
+		require.Equal(renderedDesiredSpec, specResult)
+	})
+
+	t.Run("spec from the api response has a different RenderedVersion as desired", func(t *testing.T) {
+		renderedDesiredSpec := createRenderedTestSpec(image)
+		marshaledDesiredSpec, err := json.Marshal(renderedDesiredSpec)
+		require.NoError(err)
+		mockReadWriter.EXPECT().ReadFile(desiredPath).Return(marshaledDesiredSpec, nil)
+
+		rollbackSpec, err := createEmptyTestSpec()
+		require.NoError(err)
+		mockReadWriter.EXPECT().ReadFile(rollbackPath).Return(rollbackSpec, nil)
+
+		// Api is returning a renderd version that is different from the read desired spec
+		apiResponse := &v1alpha1.RenderedDeviceSpec{RenderedVersion: "5"}
+		mockClient.EXPECT().GetRenderedDeviceSpec(ctx, gomock.Any(), gomock.Any()).Return(apiResponse, 200, nil)
+
+		// The difference results in a write call for the desired spec
```

Comment:
:+1:  Because the rendered version has changed we persist the desired state to disk then begin to reconcile.